File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.java
/*
 * @lc app=leetcode id=105 lang=java
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

// @lc code=start

import java.util.Map;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] preorder;
    private int[] inorder;
    private Map<Integer, Integer> nodeToIdxInOrder;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder = preorder;
        this.inorder = inorder;
        nodeToIdxInOrder = new HashMap<>(inorder.length);

        for(int i = 0; i < inorder.length; i++) {
            nodeToIdxInOrder.put(inorder[i], i);
        }

        return build(0, 0, inorder.length - 1);
    }

    private TreeNode build(int preStart, int inStart, int inEnd) {
        if (inStart > inEnd) {
            return null;
        }

        TreeNode root = new TreeNode(preorder[preStart]);
        int inIndex = nodeToIdxInOrder.get(root.val); 

        int leftTreeSize = inIndex - inStart;

        root.left = build(preStart + 1, inStart, inIndex - 1);
        root.right = build(preStart + leftTreeSize + 1, inIndex + 1, inEnd);

        return root;
    }
}

// @lc code=end

File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.py
#
# @lc app=leetcode id=105 lang=python3
#
# [105] Construct Binary Tree from Preorder and Inorder Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        root_idx_inorder = {}
        node_count = len(preorder)

        for idx, val in enumerate(inorder):
            root_idx_inorder[val] = idx
        
        return self.build(preorder, 0, node_count - 1, inorder, 0, node_count - 1, root_idx_inorder)
    
    def build(self, preorder, pre_start, pre_end, inorder, in_start, in_end, root_idx_inorder):
        
        if pre_start > pre_end or in_start > in_end:
            return None
        
        root = TreeNode(preorder[pre_start])
        root_idx = root_idx_inorder[root.val]
        num_left = root_idx - in_start

        root.left = self.build(preorder, pre_start + 1, pre_start + num_left, inorder, in_start, root_idx - 1, root_idx_inorder)
        root.right = self.build(preorder, pre_start + num_left + 1, pre_end, inorder, root_idx + 1, in_end, root_idx_inorder)

        return root

        



        
# @lc code=end

